tests/integration/test_prometheus_scrape_multiunit.py

```diff
@@ -0,0 +1,188 @@
+
+This test scaling up/down both sides of the relation, and upgrading.
+
+1. Deploy several units of prometheus and several units of a "provider" charm, and relate them.
+2. Confirm all units of prometheus have the correct and same targets and rules.
+3. Upgrade prometheus.
+4. Scale prometheus up and down.
+5. Scale the "provider" charm up and down.
+"""
+
+import asyncio
+import logging
+
+import pytest
+from deepdiff import DeepDiff
+from helpers import check_prometheus_is_ready, get_prometheus_active_targets, oci_image
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+prometheus_app_name = "prometheus"
+prometheus_resources = {"prometheus-image": oci_image("./metadata.yaml", "prometheus-image")}
+tester_app_name = "tester"
+num_units = 2  # Using the same number of units for both prometheus and the tester
+
+
+async def test_setup_env(ops_test: OpsTest):
+    await ops_test.model.set_config({"logging-config": "<root>=WARNING; unit=DEBUG"})
+
+
+@pytest.mark.abort_on_fail
+async def test_prometheus_scrape_relation_with_prometheus_tester(
+    ops_test: OpsTest, prometheus_charm, prometheus_tester_charm
+):
+    """Relate several units of prometheus and several units of the tester charm.
+
+    - Deploy several units of prometheus and several units of a "provider" charm, and relate them.
+    - Confirm all units of prometheus have the correct and same targets and rules.
+    """
+    app_names = [prometheus_app_name, tester_app_name]
+
+    # GIVEN prometheus and the tester charm are deployed with two units each
+
+    await asyncio.gather(
+        ops_test.model.deploy(
+            prometheus_charm,
+            resources=prometheus_resources,
+            application_name=prometheus_app_name,
+            num_units=num_units,
+        ),
+        ops_test.model.deploy(
+            prometheus_tester_charm,
+            resources={
+                "prometheus-tester-image": oci_image(
+                    "./tests/integration/prometheus-tester/metadata.yaml",
+                    "prometheus-tester-image",
+                )
+            },
+            application_name=tester_app_name,
+            num_units=num_units,
+        ),
+    )
+
+    await ops_test.model.wait_for_idle(apps=app_names, status="active", wait_for_units=num_units)
+    await asyncio.gather(
+        *[check_prometheus_is_ready(ops_test, prometheus_app_name, u) for u in range(num_units)]
+    )
+
+    # WHEN prometheus is not related to anything
+    # THEN all prometheus units should have only one scrape target (self-scraping)
+    for unit_num in range(num_units):
+        targets = await get_prometheus_active_targets(ops_test, prometheus_app_name, unit_num)
+        assert len(targets) == 1
+        self_scrape = next(iter(targets))
+        assert self_scrape["labels"]["job"] == "prometheus"
+        assert self_scrape["labels"]["instance"] == "localhost:9090"
```

Comment:
This can/should be tidied up into a single in-file helper and/or `assert all(...)`, but not really important for now.

Reply:
Yes but then the assertion failures would be too vague.

Reply:
Regardless of how vague the assertions are, a failure would require manual investigation or fixing the problematic patch anyway.

Especially considering that we don't actually indicate which unit failed, and that Prometheus self-scrape should _always_ be there. If we're really worried about it, make it `assert all(...), "No self-scrape jobs present!"`

Reply:
A fresh example: comparison of configuration failed with
```
    assert all(config_by_unit[0] == config_by_unit[u] for u in range(1, num_units))
AssertionError: assert False
 +  where False = all(<generator object test_prometheus_scrape_relation_with_prometheus_tester.<locals>.<genexpr> at 0x7f35547184a0>)
```
and the more explicit form would allow skipping a step in future failures, as the diff itself is printed.